src/mem/arena.rs

```diff
@@ -66,15 +110,80 @@ pub trait Arena {
     /// # let mut arena = BumpArena::new(&mut data);
     /// // If the first alloc succeeds (regardless of the value of `len`), then
     /// // the subsequent alloc after resetting the arena must also succeed.
-    /// assert!(arena.alloc(64).is_ok());
+    /// assert!(arena.alloc_aligned(64, 1).is_ok());
     /// arena.reset();
-    /// assert!(arena.alloc(64).is_ok());
+    /// assert!(arena.alloc_aligned(64, 1).is_ok());
     /// ```
     fn reset(&mut self);
 }
 
 assert_obj_safe!(Arena);
 
+/// Convenience functions for arenas, exposed as a trait.
+///
+/// Note that this trait is implemened for `&impl Arena`, which is the reason
+/// for the slightly odd signature.
+pub trait ArenaExt<'arena> {
+    /// Allocates a value of type `T`.
+    ///
+    /// Because of the `reset()` function, it needs to be safe to transmute
+    /// `T` back into bytes. As such, There is an additional `AsBytes` bound.
+    /// To avoid having to mess around with destructors, we additionally
+    /// require a `Copy` bound, though this may eventually be removed.
+    ///
+    /// # Panics
+    ///
+    /// This function is permitted to panic under catastrophic failure
+    /// conditions, such as completely running out of program memory.
+    /// Implementations must advertize whether they panic.
+    fn alloc<T>(self) -> Result<&'arena mut T, OutOfMemory>
+    where
+        T: AsBytes + FromBytes + Copy;
+
+    /// Allocates a slice with `n` elements.
+    ///
+    /// Because of the `reset()` function, it needs to be safe to transmute
+    /// `T` back into bytes. As such, There is an additional `AsBytes` bound.
+    /// To avoid having to mess around with destructors, we additionally
+    /// require a `Copy` bound, though this may eventually be removed.
+    ///
+    /// # Panics
+    ///
+    /// This function is permitted to panic under catastrophic failure
+    /// conditions, such as completely running out of program memory.
+    /// Implementations must advertize whether they panic.
+    fn alloc_slice<T>(self, n: usize) -> Result<&'arena mut [T], OutOfMemory>
+    where
+        T: AsBytes + FromBytes + Copy;
+}
+
+impl<'arena, A: Arena + ?Sized> ArenaExt<'arena> for &'arena A {
+    fn alloc<T: AsBytes + FromBytes + Copy>(
+        self,
+    ) -> Result<&'arena mut T, OutOfMemory> {
+        let bytes =
+            self.alloc_aligned(mem::size_of::<T>(), mem::align_of::<T>())?;
+
+        let lv = LayoutVerified::<_, T>::new(bytes)
+            .expect("alloc_aligned() implemented incorrectly");
+        Ok(lv.into_mut())
+    }
+
+    fn alloc_slice<T: AsBytes + FromBytes + Copy>(
```

Comment:
It's unclear to me whether `ArenaExt::alloc_slice` will panic, as is required by the trait method documentation (line 154).

Reply:
It can panic due to calls of alloc_aligned.

Reply:
Sorry, I worded my comment poorly. What is meant was whether it panics must be documented.